### PR TITLE
Adding shard, project, and domain labels to crd

### PR DESCRIPTION
**Pull request by hamersaw**
# TL;DR
Adding shard, project, and domain labels to the FlyteWorkflow CRD when built. These labels will be used as selectors for sharding FlytePropeller instances. The project and domain labels are self explanatory. Shard is computed as a hash of the project, domain, and workflow execution ID and modded with the keyspace size, in the initial implementation this will be 32.

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [X] Smoke tested
 - [X] Unit tests added
 - [X] Code documentation added
 - [X] Any pending items have an associated Issue

## Complete description
^^^

## Tracking Issue
_NA_

## Follow-up issue
_NA_

### CODE REVIEW

File: pkg/compiler/test/testdata/branch/k8s/mycereal_condition_has_no_deps.json [hunk 1]
```diff
@@ -6,7 +6,10 @@
     "namespace": "namespace",
     "creationTimestamp": null,
     "labels": {
+      "domain": "domain",
       "execution-id": "name",
+      "project": "hello",
+      "shard": "6",
```
Comment (kumare3) [inline pkg/compiler/test/testdata/branch/k8s/mycereal_condition_has_no_deps.json:12]: this is not really shard right? Shard is logical, it is more like a bin? Or I guess we should crisply define the semantics

File: pkg/compiler/transformers/k8s/workflow.go [hunk 2]
```diff
@@ -13,7 +14,10 @@ import (
 	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+const DomainLabel = "domain"
 const ExecutionIDLabel = "execution-id"
+const ProjectLabel = "project"
+const ShardLabel = "shard"
```
Comment (kumare3) [inline pkg/compiler/transformers/k8s/workflow.go:20]: lets write docs on this

File: pkg/compiler/transformers/k8s/workflow.go [hunk 3]
```diff
@@ -207,14 +207,20 @@ func BuildFlyteWorkflow(wfClosure *core.CompiledWorkflowClosure, inputs *core.Li
 		NodeDefaults: v1alpha1.NodeDefaults{Interruptible: interruptible},
 	}
 
-	obj.ObjectMeta.Name, obj.ObjectMeta.GenerateName, obj.ObjectMeta.Labels[ExecutionIDLabel], err =
-		generateName(wf.GetId(), executionID)
+	obj.ObjectMeta.Name, obj.ObjectMeta.GenerateName, obj.ObjectMeta.Labels[ExecutionIDLabel], obj.ObjectMeta.Labels[ProjectLabel],
```
Comment (kumare3) [inline pkg/compiler/transformers/k8s/workflow.go:210]: ohh man this is such a weird way of writing it. I so prefer capturing it into a local variable before assigning into a map

File: pkg/compiler/transformers/k8s/workflow.go [hunk 4]
```diff
@@ -207,14 +207,20 @@ func BuildFlyteWorkflow(wfClosure *core.CompiledWorkflowClosure, inputs *core.Li
...
+		obj.ObjectMeta.Labels[DomainLabel], err = generateName(wf.GetId(), executionID)
 
 	if err != nil {
 		errs.Collect(errors.NewWorkflowBuildError(err))
 	}
 	obj.ObjectMeta.Labels[WorkflowNameLabel] = utils.SanitizeLabelValue(WorkflowNameFromID(primarySpec.ID))
 
+	h := fnv.New32a()
+	h.Write([]byte(obj.ObjectMeta.Labels[ExecutionIDLabel]))
+	hash := h.Sum32() % 32
```
Comment (kumare3) [inline pkg/compiler/transformers/k8s/workflow.go:220]: can we make 32 to be a global const if not configurable?

File: pkg/compiler/test/testdata/branch/k8s/mycereal_condition_has_no_deps.json (hunk 1 above)
Comment (hamersaw) [reply pkg/compiler/test/testdata/branch/k8s/mycereal_condition_has_no_deps.json]: how does "shard-key" sound? i'm having trouble liking shard-bin, it sounds like a collection of shards - when really we're defining shards by a collection of these values.

File: pkg/compiler/transformers/k8s/workflow.go (hunk 2 above)
Comment (hamersaw) [reply pkg/compiler/transformers/k8s/workflow.go]: done

File: pkg/compiler/transformers/k8s/workflow.go (hunk 3 above)
Comment (hamersaw) [reply pkg/compiler/transformers/k8s/workflow.go]: i thought so too, was trying to stick to how it was originally written. done.

File: pkg/compiler/transformers/k8s/workflow.go (hunk 4 above)
Comment (hamersaw) [reply pkg/compiler/transformers/k8s/workflow.go:240]: done. it was in the original sharding pr, i was unsure where to put it in this one - it will certainly move.

File: pkg/compiler/transformers/k8s/workflow.go [hunk 5]
```diff
@@ -207,14 +223,24 @@ func BuildFlyteWorkflow(wfClosure *core.CompiledWorkflowClosure, inputs *core.Li
...
+	obj.ObjectMeta.GenerateName = generatedName
+	obj.ObjectMeta.Labels[ExecutionIDLabel] = label
+	obj.ObjectMeta.Labels[ProjectLabel] = project
+	obj.ObjectMeta.Labels[DomainLabel] = domain
 	obj.ObjectMeta.Labels[WorkflowNameLabel] = utils.SanitizeLabelValue(WorkflowNameFromID(primarySpec.ID))
 
+	h := fnv.New32a()
+	h.Write([]byte(label))
+	hash := h.Sum32() % 32
+
```
Comment (EngHabu) [inline pkg/compiler/transformers/k8s/workflow.go:241]:
```suggestion
	hash := h.Sum32() % v1alpha1.ShardKeyspaceSize

```

Comment (hamersaw) [reply pkg/compiler/transformers/k8s/workflow.go]: much better!

**Review by EngHabu — APPROVED** [review]
